### ISSUE TITLE
Out of bounds access in gdf_table gather_valid_mask kernel

### ISSUE THREAD

**Issue by harrism** [rapidsai/cudf#343]
I discovered this bug by running `cuda-memcheck ipython Mortgage_ETL.py`, where `Mortgage_ETL.py` is a python file exported from one of our demo notebooks. 

In `gdf_table.cuh`, the kernel `gather_valid_mask` is casting the 8-bit mask pointer to a 32-bit type

https://github.com/rapidsai/cudf/blob/612ca60a7f41adb227ad556a91c76b69d2527d9b/libgdf/src/gdf_table.cuh#L178

and then performing an `atomicOr` to replace individual bits. 

https://github.com/rapidsai/cudf/blob/612ca60a7f41adb227ad556a91c76b69d2527d9b/libgdf/src/gdf_table.cuh#L209

When `num_rows` is not a multiple of 32, this overruns the end of the array, and `cuda_memcheck` detects an invalid (out of bounds) store.

I would not be surprised if a similar bug exists in `scatter_valid_mask`.

The code needs to be modified to special case the last word of the mask when the size is not a multiple of 32. However, there is a better way to write this function.  Instead of *every thread* doing an atomicOr for every row (expensive, since you have 32-way collisions on every 32-bit word in the output mask), use `__ballot_sync()` to get the valid bits from each thread of a warp, and then have one thread of the warp (the one where `0 == (threadIdx.x % 32)) simply write the ballot result to the appropriate array location. No atomics needed, and 1/32 the memory traffic.

But the last word still needs to be handled carefully -- if it is less than 32 bits you need to split it into bytes and write those.

**Comment by harrism** [comment]
For some reason I can't assign you @kaatish, but it looks like you wrote this code?

**Comment by harrism** [comment]
Oops, look like @jrhemstad wrote it. Sorry @kaatish.

**Comment by jrhemstad** [comment]
This can be fixed once https://github.com/rapidsai/cudf/pull/576 is merged.

**Comment by jrhemstad** [comment]
Fixed by https://github.com/rapidsai/cudf/pull/1153
